documentation/bootstrap_upgrade_notes.markdown

```diff
@@ -4,25 +4,35 @@ title: "Puppet Server: Bootstrap Upgrade Notes"
 
-Starting in Puppet Server 2.5.0 this problem will start to be resolved. However, it's important to note that users who have disabled the CA service in bootstrap.cfg will need to exercise caution when upgrading to 2.5.0. Subsequent releases should no longer be subject to this issue. This page will walk you through what upgrading will be like if you are managing `bootstrap.cfg` yourself.
+Users of Puppet Server 2.4.x and earlier could modify their [`bootstrap.cfg` file][] in order to disable the CA on compile masters and support a multi-master configuration. Upgrades between these older versions have been painful, however, due to package managers attempting to overwrite this file during upgrades.
+
+This could cause two problems:
+
+1.  If users disabled CA services and chose the packaged version during the upgrade, CA services would be re-enabled on the master after the upgrade, which could break their multi-master setup.
+
+2.  If users disabled CA services and chose their version of `bootstrap.cfg`, and the new version contained settings for new services that were added to the packaged version of `bootstrap.cfg`, and in that case, the server will fail to start.
+
+Puppet Server 2.5.0 takes the first steps toward resolving this problem while maintaining configurability by changing how service bootstrap configuration works. However, users of Puppet Server 2.4.x and older who disabled the CA service in `bootstrap.cfg` **must take special precautions when upgrading to 2.5.0** to prevent the upgrade process from re-enabling the CA service or potentially overwriting files in the `ssldir`. (Subsequent releases should no longer be subject to this issue.)
 
 ## Upgrading to 2.5.0 or newer
 
-If you are not managing your `bootstrap.cfg` file, you don't need to do anything.
+Puppet Server 2.5.0 no longer uses the [`bootstrap.cfg` file][] to configure service bootstrapping. Instead, it reads files within the `/etc/puppetlabs/puppetserver/services.d/` directory, which can contain multiple files --- some designed to be edited by users --- that configure service bootstrapping.
+
+If you edited or manage your `bootstrap.cfg` file, do the following:
 
-If you are managing your `bootstrap.cfg`, here's what you need to be aware of:
+### Before you upgrade: `ca.cfg`
 
-### `ca.cfg`
+> **Warning:** Back up your masters' [`ssldir`](https://docs.puppet.com/puppet/latest/reference/dirs_ssldir.html) (or at least your `crl.pem` file) before the upgrade. If anything unexpected happens, such as a master's newly enabled CA services or an emergency rollback overwriting them during the upgrade, this ensures that you can restore your previous certificates and certificate revocation list.
```

Comment:
`If anything unexpected happens, such as a master's newly enabled CA services or an emergency rollback overwriting them during the upgrade` - that part feels like it needs a tweak or two.